ssl/ech/ech_internal.c

```diff
@@ -160,4 +216,1107 @@ int ossl_ech_conn_init(SSL_CONNECTION *s, SSL_CTX *ctx,
+    if (buf != NULL && blen > 0)
+        BIO_write(s->s3.handshake_buffer, (void *)buf, (int)blen);
+    return 1;
+}
+
+/*
+ * To control the number of zeros added after an EncodedClientHello - we pad
+ * to a target number of octets or, if there are naturally more, to a number
+ * divisible by the defined increment (we also do the spec-recommended SNI
+ * padding thing first)
+ */
+# define OSSL_ECH_PADDING_TARGET 128 /* ECH cleartext padded to at least this */
+# define OSSL_ECH_PADDING_INCREMENT 32 /* ECH padded to a multiple of this */
+
+/*
+ * figure out how much padding for cleartext (on client)
+ * ee is the chosen ECHConfig
+ * return overall length to use including padding or zero on error
+ *
+ * "Recommended" inner SNI padding scheme as per spec (section 6.1.3)
+ * Might remove the mnl stuff later - overall message padding seems
+ * better really, BUT... we might want to keep this if others (e.g.
+ * browsers) do it so as to not stand out compared to them.
+ *
+ * The "+ 9" constant below is from the specifiation and is the
+ * expansion comparing a string length to an encoded SNI extension.
+ * Same is true of the 31/32 formula below.
+ *
+ * Note that the AEAD tag will be added later, so if we e.g. have
+ * a padded cleartext of 128 octets, the ciphertext will be 144
+ * octets.
+ */
+static size_t ech_calc_padding(SSL_CONNECTION *s, OSSL_ECHSTORE_ENTRY *ee,
+                               size_t encoded_len)
+{
+    int length_of_padding = 0, length_with_snipadding = 0;
+    int innersnipadding = 0, length_with_padding = 0;
+    size_t mnl = 0, clear_len = 0, isnilen = 0;
+
+    if (s == NULL || ee == NULL)
+        return 0;
+    mnl = ee->max_name_length;
+    if (mnl != 0) {
+        /* do weirder padding if SNI present in inner */
+        if (s->ext.hostname != NULL) {
+            isnilen = strlen(s->ext.hostname) + 9;
+            innersnipadding = (mnl > isnilen) ? mnl - isnilen : 0;
+        } else {
+            innersnipadding = mnl + 9;
+        }
+    }
+    /* padding is after the inner client hello has been encoded */
+    length_with_snipadding = innersnipadding + encoded_len;
+    length_of_padding = 31 - ((length_with_snipadding - 1) % 32);
+    length_with_padding = encoded_len + length_of_padding + innersnipadding;
+    /*
+     * Finally - make sure final result is longer than padding target
+     * and a multiple of our padding increment.
+     * TODO(ECH): This is a local addition - we might take it out if
+     * it makes us stick out; or if we take out the above more (uselessly:-)
+     * complicated scheme, we may only need this in the end.
+     */
+    if (length_with_padding % OSSL_ECH_PADDING_INCREMENT)
+        length_with_padding += OSSL_ECH_PADDING_INCREMENT
+            - (length_with_padding % OSSL_ECH_PADDING_INCREMENT);
+    while (length_with_padding < OSSL_ECH_PADDING_TARGET)
+        length_with_padding += OSSL_ECH_PADDING_INCREMENT;
+    clear_len = length_with_padding;
```

Comment:
Not sure why we do this assignment. We don't seem to do anything with it, other than return it. Why not just return `length_with_padding`?

Reply:
Yeah `clear_len` is a bit of a hangover from some earlier (near forgotten:-) padding scheme where it could've differed from `length_with_padding`.  I'll ditch `clear_len`.